chans/chans_test.go

```diff
@@ -70,9 +73,42 @@ func TestFlatten(t *testing.T) {
 	}()
 
 	for i := 0; i < 99; i += 3 {
-		source <- []int{i, i + 1, i + 2}
+		c := make(chan int, 3)
+		source <- c
+		c <- i
+		c <- i + 1
+		c <- i + 2
 	}
 	close(source)
 	wg.Wait()
 	assert.Equal(t, 4851, total)
 }
+
+func TestPushPop(t *testing.T) {
+	t.Run("valid ctx", func(t *testing.T) {
+		c := make(chan int, 1)
+		assert.True(t, chans.Push(context.Background(), c, 7))
+		i, ok := chans.Pop(context.Background(), c)
+		assert.True(t, ok)
+		assert.Equal(t, 7, i)
+	})
+
+	t.Run("cancelled ctx", func(t *testing.T) {
+		c := make(chan int, 1)
+		cancelledCtx, cancel := context.WithCancel(context.Background())
+		cancel()
+		assert.False(t, chans.Push(cancelledCtx, c, 7))
+		_, ok := chans.Pop(cancelledCtx, c)
+		assert.False(t, ok)
+	})
+
+	t.Run("cancel ctx while blocked", func(t *testing.T) {
+		c := make(chan int)
+		cancelledCtx, cancel := context.WithTimeout(context.Background(), 50*time.Millisecond)
+		defer cancel()
+
```

Comment:
Do you not want to sleep here for 50ms to ensure that the ctx has timed out?

Reply:
no, the point is to get into the Push and Pop functions before the context has timed out and only then it expires. I added a second ctx for the Pop though